fl4health/checkpointing/opacus_checkpointer.py

```diff
@@ -0,0 +1,167 @@
+import pickle
+from logging import INFO
+from typing import Any, Dict
+
+import torch.nn as nn
+from flwr.common.logger import log
+from flwr.common.typing import Scalar
+from opacus import GradSampleModule
+
+from fl4health.checkpointing.checkpointer import FunctionTorchCheckpointer
+
+
+class OpacusCheckpointer(FunctionTorchCheckpointer):
+    """
+    This is a specific type of checkpointer to be used in saving models trained using Opacus for differential privacy.
+    Certain layers within Opacus wrapped models do not interact well with torch.save functionality. This checkpointer
+    fixes this issue.
+    """
+
+    def maybe_checkpoint(self, model: nn.Module, loss: float, metrics: Dict[str, Scalar]) -> None:
+        """
+        Overriding the checkpointing strategy of the FunctionTorchCheckpointer to save model state dictionaries
+        instead of using the torch.save workflow.
+
+        Args:
+            model (nn.Module): Model to be potentially saved (should be an Opacus wrapped model)
+            loss (float): Loss value associated with the model to be used in checkpointing decisions.
+            metrics (Dict[str, Scalar]): Metrics associated with the model to be used in checkpointing decisions.
+        """
+        assert isinstance(
+            model, GradSampleModule
+        ), f"Model is of type: {type(model)}. This checkpointer need only be used to checkpoint Opacus modules"
```

Comment:
Used to force the use of this checkpointer only for Opacus based models.

Reply:
There is no way of changing the type in the method signature?

Reply:
You are correct, we actually can change the signature here. So I did that. However, I still want to only allow this checkpointer to be used for Opacus models. If the library user isn't using mypy, they'll still be able to call maybe_checkpoint. So I left the assertions in there, but changed the signature.